### PR TITLE
Python: Filter on Datafile metrics

### LINKED ISSUE #6567: Python: Table scan problem with row filter set to non-partition columns
### Apache Iceberg version

None

### Query engine

None

### Please describe the bug 🐞

I really like the new table scan feature released latest pyiceberg 0.2.1 release, thanks @Fokko. It works great when I provide the partition column as row filter but not working as expected when I provide other columns as part of expression. The `scan.plan_files()` shall return me collection of parquet files satisfy the predicate in row filter but it's returning all instead. 

Here's my repro steps, I created a simple table `hongyue_zhang.mls23` to start with

### schema
```ddl
                                       Create Table
------------------------------------------------------------------------------------------
 CREATE TABLE iceberg.hongyue_zhang.mls23 (
    id bigint NOT NULL,
    data varchar,
    ts date
 )
 WITH (
    format = 'PARQUET',
    location = 's3a://warehouse-default/warehouse/hongyue_zhang.db/mls23',
    partitioning = ARRAY['ts']
 )
(1 row)
```

### Setup
Table have 2 partitions  and 198 records total, each write have its own parquet files for the sake of simplicity 

```sql
-- loop through 
f"INSERT INTO TABLE {catalog}.{table_name} VALUES ({id}, '{data}', DATE '{str(ts)}');"
```
```
     partition    | record_count | file_count | total_size |                                        data
-----------------+--------------+------------+------------+-------------------------------------------------------------------------------------
 {ts=2023-01-04} |           99 |         99 |     115300 | {id={min=1, max=1, null_count=0}, data={min=b, max=bbbbbbbbbbbbbbbc, null_count=0}}
 {ts=2023-01-05} |           99 |         99 |     115303 | {id={min=0, max=0, null_count=0}, data={min=a, max=aaaaaaaaaaaaaaab, null_count=0}}
(2 rows)
```

### Python code
```python
import os
from pyiceberg.catalog import load_catalog
from pyiceberg.expressions import GreaterThanOrEqual, And, EqualTo

catalog = load_catalog("prod")
table = catalog.load_table("hongyue_zhang.mls23")
table.location()

scan1 = table.scan(
    row_filter=EqualTo("ts", "2023-01-04"))
yesterday_files = [task.file.file_path for task in scan1.plan_files()]
print(len(yesterday_files))
# expect 99 and actual is 99 parquet files for single partition

scan2 = table.scan(
    row_filter=EqualTo("data", "a"))
a_files = [task.file.file_path for task in scan2.plan_files()]
print(len(a_files))
# expect 1 but I am seeing 198 instead, which means all parquet files are returned

scan3 = table.scan(
    row_filter=And(EqualTo("ts", "2023-01-04"), EqualTo("data", "a")))
yesterday_and_a_files= [task.file.file_path for task in scan3.plan_files()]
print(len(yesterday_and_a_files))
# expect 1 but I am seeing 99, which means the row filter are taking the 1st expression with partition column ts but not 2nd expression on data 
```

For the sake of validation, I also tried to spark to query with similar condition and it's returnning me 1 file as expected
```spark
val result = spark.sql("select id, data, input_file_name(), ts from iceberg.hongyue_zhang.mls23 where data = 'a'")
+---+----+-----------------------------------------------------------------------------------------------------------------------------------------------------+----------+
|id |data|input_file_name()                                                                                                                                    |ts        |
+---+----+-----------------------------------------------------------------------------------------------------------------------------------------------------+----------+
|0  |a   |s3a://warehouse-default/warehouse/hongyue_zhang.db/mls23/data/ts=2023-01-05/00000-6-5573682f-d72c-4a68-a08f-8fe4dbca8581-00001.parquet               |2023-01-05|
+---+----+-----------------------------------------------------------------------------------------------------------------------------------------------------+----------+
```

I cant see to figure out why it failed and happy to contribute if anyone can share insights

### PR DESCRIPTION
**Pull request by Fokko**
This allows us to skip full data files.

When it hits data:
```python
➜  python git:(master) ✗ python3
Python 3.10.9 (main, Dec 15 2022, 17:11:09) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyiceberg.catalog import load_catalog
>>> 
>>> catalog = load_catalog('local')
>>> 
>>> tbl = catalog.load_table('nyc.taxis')
>>> 
>>> from pyiceberg.expressions import EqualTo
>>> 
>>> sc = tbl.scan(row_filter=EqualTo("PULocationID", "151")).to_arrow()
>>> 
>>> sc.to_pandas()
        VendorID      tpep_pickup_datetime     tpep_dropoff_datetime  passenger_count  trip_distance  ...  tolls_amount improvement_surcharge  total_amount  congestion_surcharge  airport_fee
0              2 2021-12-01 00:03:00+00:00 2021-12-01 00:40:00+00:00              NaN          13.87  ...          6.55                   0.3         43.83                   NaN          NaN
1              2 2021-12-01 02:05:40+00:00 2021-12-01 02:19:18+00:00              NaN           4.14  ...          0.00                   0.3         24.01                   NaN          NaN
2              2 2021-12-01 04:37:01+00:00 2021-12-01 04:47:40+00:00              NaN           5.23  ...          0.00                   0.3         25.39                   NaN          NaN
3              2 2021-12-01 00:39:29+00:00 2021-12-01 00:43:15+00:00              1.0           0.81  ...          0.00                   0.3         10.56                   2.5          0.0
4              2 2021-12-01 04:48:00+00:00 2021-12-01 05:00:00+00:00              NaN           3.46  ...          0.00                   0.3         23.04                   NaN          NaN
...          ...                       ...                       ...              ...            ...  ...           ...                   ...           ...                   ...          ...
155198         1 2022-04-30 23:57:28+00:00 2022-05-01 00:11:03+00:00              NaN           0.00  ...          0.00                   0.3         19.25                   NaN          NaN
155199         2 2022-04-30 23:44:15+00:00 2022-04-30 23:55:07+00:00              NaN           2.77  ...          0.00                   0.3         19.61                   NaN          NaN
155200         2 2022-04-30 23:05:00+00:00 2022-04-30 23:17:00+00:00              NaN           4.57  ...          0.00                   0.3         26.44                   NaN          NaN
155201         2 2022-04-30 23:10:33+00:00 2022-04-30 23:36:09+00:00              NaN          11.38  ...          6.55                   0.3         57.38                   NaN          NaN
155202         2 2022-04-30 23:00:00+00:00 2022-04-30 23:26:00+00:00              NaN           6.76  ...          0.00                   0.3         38.57                   NaN          NaN

[155203 rows x 19 columns]
```

Or when it doesn't:
```python
➜  python git:(master) ✗ python3
Python 3.10.9 (main, Dec 15 2022, 17:11:09) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from pyiceberg.catalog import load_catalog
>>> 
>>> catalog = load_catalog('local')
>>> 
>>> tbl = catalog.load_table('nyc.taxis')
>>> 
>>> from pyiceberg.expressions import EqualTo
>>> 
>>> sc = tbl.scan(row_filter=EqualTo("PULocationID", "1000")).to_arrow()
>>> 
>>> sc
pyarrow.Table
VendorID: int64
tpep_pickup_datetime: timestamp[us, tz=+00:00]
tpep_dropoff_datetime: timestamp[us, tz=+00:00]
passenger_count: double
trip_distance: double
RatecodeID: double
store_and_fwd_flag: string
PULocationID: int64
DOLocationID: int64
payment_type: int64
fare_amount: double
extra: double
mta_tax: double
tip_amount: double
tolls_amount: double
improvement_surcharge: double
total_amount: double
congestion_surcharge: double
airport_fee: double
----
VendorID: []
tpep_pickup_datetime: []
tpep_dropoff_datetime: []
passenger_count: []
trip_distance: []
RatecodeID: []
store_and_fwd_flag: []
PULocationID: []
DOLocationID: []
payment_type: []
...
```

Fixes #6567 

cc @dramaticlly

### CODE REVIEW

File: python/pyiceberg/expressions/visitors.py [hunk 1]
```diff
@@ -986,3 +989,246 @@ def expression_to_plain_format(
...
+        self.value_counts = file.value_counts or EMPTY_DICT
+        self.null_counts = file.null_value_counts or EMPTY_DICT
+        self.nan_counts = file.nan_value_counts or EMPTY_DICT
+        self.lower_bounds = file.lower_bounds or EMPTY_DICT
+        self.upper_bounds = file.upper_bounds or EMPTY_DICT
+
+        return visit(self.expr, self)
+
+    def _contains_nulls_only(self, idx: int) -> bool:
+        return idx in self.value_counts and idx in self.null_counts and self.value_counts[idx] - self.null_counts[idx] == 0
```
Comment (dramaticlly) [inline python/pyiceberg/expressions/visitors.py:1030]: I guess column index for value and null count always present even when count = 0 , but not necessarily to have idx=0 for nan count.

File: python/tests/expressions/test_visitors.py [hunk 2]
```diff
@@ -1472,3 +1478,211 @@ def test_dnf_to_dask(table_schema_simple: Schema) -> None:
...
+    assert should_read, "Should match: no visibility"
+
+    should_read = _InclusiveMetricsEvaluator(schema_data_file, In("all_nan_null_bounds", (1, 10, 30))).eval(data_file)
+    assert not should_read, "Should not match: all nan column doesn't contain number"
+
+    should_read = _InclusiveMetricsEvaluator(schema_data_file, In("some_nan_correct_bounds", (1, 10, 30))).eval(data_file)
+    assert should_read, "Should match: 10 within bounds"
+
+    should_read = _InclusiveMetricsEvaluator(schema_data_file, In("some_nan_correct_bounds", (1, 30))).eval(data_file)
+    assert not should_read, "Should not match: 1 not within bounds"
```
Comment (dramaticlly) [inline python/tests/expressions/test_visitors.py:1659]: nit: Should not match, 1 and 30 are both are out of bounds

**Review by dramaticlly — COMMENTED** [review]
Thank you so much @Fokko , love this! 🚀

File: python/pyiceberg/expressions/literals.py [hunk 3]
```diff
@@ -328,6 +328,9 @@ def __le__(self, other: Any) -> bool:
     def __ge__(self, other: Any) -> bool:
         return self._value32 >= other
 
+    def __hash__(self) -> int:
+        return hash(self.value)
```
[the remark this answers is not in the document]
Comment (Fokko) [reply python/pyiceberg/expressions/literals.py:332]: They are identical:
```
➜  python git:(fd-filter-on-statistics) python3
Python 3.10.9 (main, Dec 15 2022, 17:11:09) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import struct
>>> 
>>> hash(struct.unpack("<f", struct.pack("<f", 22))[0])
22
>>> hash(22)
22
>>> struct.unpack("<f", struct.pack("<f", 22))[0]
22.0
>>> 22
22
>>> 22.0
22.0
```
I'm not sure why we have the `_value32` there anyway. Looks like we're just packing and unpacking the value.

File: python/tests/expressions/test_visitors.py (hunk 2 above)
Comment (Fokko) [reply python/tests/expressions/test_visitors.py:1659]: Good call, just copied this from Java :)

File: python/pyiceberg/expressions/visitors.py (hunk 1 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1030]: The dictionary can be empty if I'm not mistaking

**Review by dramaticlly — APPROVED** [review]
Thank you @Fokko , cant wait to use this in 0.4.0 release!

File: python/pyiceberg/expressions/visitors.py [hunk 4]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+
+        self.value_counts = file.value_counts or EMPTY_DICT
+        self.null_counts = file.null_value_counts or EMPTY_DICT
+        self.nan_counts = file.nan_value_counts or EMPTY_DICT
+        self.lower_bounds = file.lower_bounds or EMPTY_DICT
+        self.upper_bounds = file.upper_bounds or EMPTY_DICT
+
+        return visit(self.expr, self)
+
+    def _contains_nulls_only(self, idx: int) -> bool:
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1028]: These are field IDs, not indexes, right? I think it would make more sense to use `id` instead since many people confuse the IDs with position indexes.

File: python/pyiceberg/expressions/visitors.py [hunk 5]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+    def eval(self, file: DataFile) -> bool:
+        """Test whether the file may contain records that match the expression."""
+
+        if file.record_count == 0:
+            return ROWS_CANNOT_MATCH
+
+        if file.record_count < 0:
+            # @TODO we haven't implemented parsing record count from avro file and thus set record count -1
+            # when importing avro tables to iceberg tables. This should be updated once we implemented
+            # and set correct record count.
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1017]: I think that we have implemented correct value counts for Avro tables, so we don't need a TODO. Instead, we should say that some older versions may have set this to -1 and that we don't trust stats in that case so we must return that rows might match.

File: python/pyiceberg/expressions/visitors.py [hunk 6]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+        self.lower_bounds = file.lower_bounds or EMPTY_DICT
+        self.upper_bounds = file.upper_bounds or EMPTY_DICT
+
+        return visit(self.expr, self)
+
+    def _contains_nulls_only(self, idx: int) -> bool:
+        return idx in self.value_counts and idx in self.null_counts and self.value_counts[idx] - self.null_counts[idx] == 0
+
+    def _contains_nans_only(self, idx: int) -> bool:
+        return idx in self.nan_counts and idx in self.value_counts and self.nan_counts[idx] == self.value_counts[idx]
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1032]: This uses `==` but the check above for nulls uses `(count - null_count) == 0`. Not a big deal, but consistency is nice.

File: python/pyiceberg/expressions/visitors.py [hunk 7]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+    def visit_and(self, left_result: bool, right_result: bool) -> bool:
+        return left_result and right_result
+
+    def visit_or(self, left_result: bool, right_result: bool) -> bool:
+        return left_result or right_result
+
+    def visit_is_null(self, term: BoundTerm[L]) -> bool:
+        idx = term.ref().field.field_id
+
+        if self.null_counts is not None and idx in self.null_counts and self.null_counts[idx] == 0:
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1057]: Other places assume that `null_counts` is non-null. Do we need to check this here?

File: python/pyiceberg/expressions/visitors.py [hunk 8]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+
+    def visit_less_than(self, term: BoundTerm[L], literal: Literal[L]) -> bool:
+        field = term.ref().field
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1101]: Is this check needed? At this point, the expression is bound and we know that the literal and the term match. I don't think that we support any way to create a literal for a nested type or to bind one, so it seems safe to assume that this is fine.

Also, if we were to support nested types then the check here would be identical for nested values. We would deserialize the lower bound, then use the literal's comparison methods to test.

File: python/pyiceberg/expressions/visitors.py [hunk 9]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+        return visit(self.expr, self)
+
+    def _contains_nulls_only(self, idx: int) -> bool:
+        return idx in self.value_counts and idx in self.null_counts and self.value_counts[idx] - self.null_counts[idx] == 0
+
+    def _contains_nans_only(self, idx: int) -> bool:
+        return idx in self.nan_counts and idx in self.value_counts and self.nan_counts[idx] == self.value_counts[idx]
+
+    def _is_nan(self, val: Any) -> bool:
+        return math.isnan(val)
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1035]: Does this handle `None` and non-float values?

File: python/pyiceberg/expressions/visitors.py [hunk 10]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.lower_bounds:
+            lower = from_bytes(field.field_type, self.lower_bounds[idx])
+
+            if self._is_nan(lower):
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1106]: This could potentially pass `None` and non-float values. We should make sure that `_is_nan` handles those correctly.

File: python/pyiceberg/expressions/visitors.py [hunk 11]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+        idx = field.field_id
+
+        if self._contains_nulls_only(idx) or self._contains_nans_only(idx):
+            return ROWS_CANNOT_MATCH
+
+        if not isinstance(field.field_type, PrimitiveType):
+            raise ValueError(f"Expected PrimitiveType: {field.field_type}")
+
+        if idx in self.upper_bounds and from_bytes(field.field_type, self.upper_bounds[idx]) < literal.value:
+            return ROWS_CANNOT_MATCH
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1163]: Why do the greater-than implementations omit the NaN check that was used in less-than and equal implementations?

Comment (rdblue) [reply python/pyiceberg/expressions/visitors.py:1163]: Oh, I see Java does it too. My guess is that we don't expect NaN for upper bounds, but it's probably safer just to run the check everywhere.

File: python/pyiceberg/expressions/visitors.py [hunk 12]
```diff
@@ -986,3 +989,245 @@ def expression_to_plain_format(
...
+                return ROWS_MIGHT_MATCH
+
+            literals = {lit for lit in literals if lower <= lit}
+            if len(literals) == 0:
+                return ROWS_CANNOT_MATCH
+
+        if idx in self.upper_bounds:
+            upper = from_bytes(field.field_type, self.upper_bounds[idx])
+            # this is different from Java, here NaN is always larger
+            if not self._is_nan(upper):
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1223]: This seems equivalent to the `if self._is_nan(lower)` in other places. Maybe we should use the same check and return `ROWS_MIGHT_MATCH` if the bound is NaN.

File: python/pyiceberg/io/pyarrow.py [hunk 13]
```diff
@@ -555,8 +555,10 @@ def project_table(
 
     if len(tables) > 1:
         return pa.concat_tables(tables)
-    else:
+    elif len(tables) == 1:
         return tables[0]
+    else:
+        return pa.Table.from_batches([], schema=schema_to_pyarrow(projected_schema))
```
Comment (rdblue) [inline python/pyiceberg/io/pyarrow.py:561]: Should this be in a separate PR?

File: python/pyiceberg/table/__init__.py [hunk 14]
```diff
@@ -314,11 +314,19 @@ def _check_content(file: DataFile) -> DataFile:
...
+    row_filter: BooleanExpression,
+    manifest: ManifestFile,
+    partition_filter: Callable[[DataFile], bool],
+    case_sensitive: bool,
+) -> List[FileScanTask]:
     all_files = files(io.new_input(manifest.manifest_path))
     matching_partition_files = filter(partition_filter, all_files)
     matching_partition_data_files = map(_check_content, matching_partition_files)
-    return [FileScanTask(file) for file in matching_partition_data_files]
+    evaluator = _InclusiveMetricsEvaluator(schema, row_filter, case_sensitive)
```
Comment (rdblue) [inline python/pyiceberg/table/__init__.py:328]: Do we need to create a new metrics evaluator for each manifest? The table schema, row filter, and case sensitive values are always the same.

I think we could handle this like `partition_filter` and pass in a `Callable`.

File: python/tests/expressions/test_visitors.py [hunk 15]
```diff
@@ -1486,3 +1492,211 @@ def test_dnf_to_dask(table_schema_simple: Schema) -> None:
...
+        upper_bounds={
+            1: to_bytes(DoubleType(), float("nan")),
+            2: to_bytes(DoubleType(), float("nan")),
+            3: to_bytes(FloatType(), float("nan")),
+            5: to_bytes(FloatType(), 22),
+        },
+    )
+
+
+def test_inclusive_metrics_evaluator_less_than_and_less_than_equal(schema_data_file: Schema, data_file: DataFile) -> None:
```
Comment (rdblue) [inline python/tests/expressions/test_visitors.py:1553]: Minor, but I think it makes sense to keep tests like these in separate files. That way they're easier to find and run. There's not much of a downside to having tests spread across files, too.

File: python/tests/expressions/test_visitors.py [hunk 16]
```diff
@@ -1486,3 +1492,211 @@ def test_dnf_to_dask(table_schema_simple: Schema) -> None:
...
+            5: 10,
+        },
+        value_counts={
+            1: 10,
+            2: 10,
+            3: 10,
+            4: 10,
+            5: 10,
+        },
+        null_value_counts={
```
Comment (rdblue) [inline python/tests/expressions/test_visitors.py:1530]: I think we will need tests for an `all_null` column.

File: python/tests/expressions/test_visitors.py [hunk 17]
```diff
@@ -1486,3 +1492,211 @@ def test_dnf_to_dask(table_schema_simple: Schema) -> None:
...
+            1: to_bytes(DoubleType(), float("nan")),
+            2: to_bytes(DoubleType(), float("nan")),
+            3: to_bytes(FloatType(), float("nan")),
+            5: to_bytes(FloatType(), 22),
+        },
+    )
+
+
+def test_inclusive_metrics_evaluator_less_than_and_less_than_equal(schema_data_file: Schema, data_file: DataFile) -> None:
+    for operator in [LessThan, LessThanOrEqual]:
```
Comment (rdblue) [inline python/tests/expressions/test_visitors.py:1554]: Why would the tests be the same for `<` and `<=`?

Those have different boundary conditions that need to be tested.

File: python/tests/expressions/test_visitors.py [hunk 18]
```diff
@@ -1486,3 +1492,211 @@ def test_dnf_to_dask(table_schema_simple: Schema) -> None:
...
+    assert not should_read, "Should not match: 1 and 30 not within bounds"
+
+    should_read = _InclusiveMetricsEvaluator(schema_data_file, In("some_nan_correct_bounds", (5, 7))).eval(data_file)
+    assert should_read, "Should match: overlap with lower bound"
+
+    should_read = _InclusiveMetricsEvaluator(schema_data_file, In("some_nan_correct_bounds", (22, 25))).eval(data_file)
+    assert should_read, "Should match: overlap with upper bounds"
+
+
+def test_inclusive_metrics_evaluator_not_in(schema_data_file: Schema, data_file: DataFile) -> None:
```
Comment (rdblue) [inline python/tests/expressions/test_visitors.py:1682]: I don't see tests for `and` or `or`, or some other cases from the Java tests. I think we need those.

Comment (Fokko) [reply python/tests/expressions/test_visitors.py:1682]: Ah, I've ported `TestMetricsEvaluatorsNaNHandling` and missed `TestInclusiveMetricsEvaluator`. Let me port that one as well.

File: python/tests/expressions/test_visitors.py (hunk 17 above)
Comment (Fokko) [reply python/tests/expressions/test_visitors.py:1554]: This is from `TestMetricsEvaluatorsNaNHandling` that focusses on `NaN` handling, let me port `TestInclusiveMetricsEvaluator` as well 👍🏻

File: python/pyiceberg/expressions/visitors.py (hunk 6 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1032]: Copied this from Java :) I've updated it to `==`

File: python/pyiceberg/expressions/visitors.py (hunk 7 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1057]: I think we need to check if it is non-null until we get default values, updated this in the other places

File: python/pyiceberg/expressions/visitors.py (hunk 4 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1028]: `id` is a reserved keyword in Python:
```
python3
Python 3.10.10 (main, Feb  8 2023, 05:34:50) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> id("vo")
4369414704
```
I've updated this to `field_id`

File: python/pyiceberg/expressions/visitors.py (hunk 8 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1101]: This is just to make mypy happy

File: python/pyiceberg/expressions/visitors.py (hunk 9 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1035]: Good question. It throws an error:
```
>>> math.isnan(None)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: must be real number, not NoneType
>>> math.isnan("vo")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: must be real number, not str
```
I think this should be fine because a non-float/double field can never be NaN. If you do `IsNan` expression on a string field, it will just skip the file.

File: python/pyiceberg/expressions/visitors.py (hunk 10 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1106]: Good call, I've updated this to:
```python
    def _is_nan(self, val: Any) -> bool:
        try:
            return math.isnan(val)
        except TypeError:
            # In the case of None or other non-numeric types
            return False
```

File: python/pyiceberg/expressions/visitors.py (hunk 12 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1223]: Good one, updated 👍🏻

File: python/pyiceberg/io/pyarrow.py (hunk 13 above)
Comment (Fokko) [reply python/pyiceberg/io/pyarrow.py:561]: I can split this into a separate PR, but I think it makes sense to ship it with this PR since if all the datafiles are skipped, it will return an empty iterator that can't be concatenated.

File: python/pyiceberg/table/__init__.py (hunk 14 above)
Comment (Fokko) [reply python/pyiceberg/table/__init__.py:328]: Oeh, I like that _very_ much! Thanks for the sharp observation!

File: python/tests/expressions/test_visitors.py (hunk 15 above)
Comment (Fokko) [reply python/tests/expressions/test_visitors.py:1553]: Split it into `test_evaluator.py` 👍🏻

File: python/tests/expressions/test_visitors.py (hunk 16 above)
Comment (Fokko) [reply python/tests/expressions/test_visitors.py:1530]: I've copied all the Java tests that include the `all_null` column

File: python/pyiceberg/expressions/visitors.py [hunk 19]
```diff
@@ -986,3 +989,263 @@ def expression_to_plain_format(
...
+        self.nan_counts = file.nan_value_counts or EMPTY_DICT
+        self.lower_bounds = file.lower_bounds or EMPTY_DICT
+        self.upper_bounds = file.upper_bounds or EMPTY_DICT
+
+        return visit(self.expr, self)
+
+    def _contains_nulls_only(self, field_id: int) -> bool:
+        return (
+            self.value_counts is not None
+            and self.null_counts is not None
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1031]: We initialize these to EMPTY_DICT, so we shouldn't need to check whether the maps are null, right?

File: python/pyiceberg/expressions/visitors.py [hunk 20]
```diff
@@ -986,3 +989,263 @@ def expression_to_plain_format(
...
+            and field_id in self.value_counts
+            and field_id in self.null_counts
+            and self.value_counts[field_id] == self.null_counts[field_id]
+        )
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1035]: I tend to avoid multiple lookups in hashmaps. What about using `get` to avoid the `KeyError` instead?

```python
def _contains_nulls_only(self, field_id: int) -> bool:
    if count := self.value_counts.get(field_id):
        if null_count := self.null_counts.get(field_id):
            return count == null_count
    return false
```

File: python/pyiceberg/expressions/visitors.py (hunk 4 above)
Comment (rdblue) [reply python/pyiceberg/expressions/visitors.py]: +1 for `field_id`. That's much more clear.

File: python/pyiceberg/expressions/visitors.py (hunk 6 above)
Comment (rdblue) [reply python/pyiceberg/expressions/visitors.py]: Yeah, sorry about that. Probably my fault!

File: python/pyiceberg/expressions/visitors.py [hunk 21]
```diff
@@ -986,3 +989,263 @@ def expression_to_plain_format(
...
+    def visit_and(self, left_result: bool, right_result: bool) -> bool:
+        return left_result and right_result
+
+    def visit_or(self, left_result: bool, right_result: bool) -> bool:
+        return left_result or right_result
+
+    def visit_is_null(self, term: BoundTerm[L]) -> bool:
+        field_id = term.ref().field.field_id
+
+        if self.null_counts is not None and field_id in self.null_counts and self.null_counts[field_id] == 0:
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1073]: Another check for a map that is defaulted.

File: python/pyiceberg/expressions/visitors.py (hunk 11 above)
Comment (rdblue) [reply python/pyiceberg/expressions/visitors.py:1179]: @Fokko, looks like this wasn't updated. The less_than and greater_than cases still differ in structure because the null check isn't there.

File: python/pyiceberg/expressions/visitors.py [hunk 22]
```diff
@@ -986,3 +989,263 @@ def expression_to_plain_format(
...
+
+            if self._is_nan(lower):
+                # NaN indicates unreliable bounds. See the InclusiveMetricsEvaluator docs for more.
+                return ROWS_MIGHT_MATCH
+
+            if lower > literal.value:
+                return ROWS_CANNOT_MATCH
+
+        if field_id in self.upper_bounds and from_bytes(field.field_type, self.upper_bounds[field_id]) < literal.value:
+            return ROWS_CANNOT_MATCH
```
Comment (rdblue) [inline python/pyiceberg/expressions/visitors.py:1204]: I think we want to match the structure of the lower bounds check and include an `isnan` check here, too.

File: python/pyiceberg/expressions/visitors.py (hunk 12 above)
Comment (rdblue) [reply python/pyiceberg/expressions/visitors.py]: Looks like this was updated for `visit_in` but not `visit_equal`

File: python/tests/expressions/test_visitors.py [hunk 23]
```diff
@@ -1024,7 +1024,7 @@ def test_missing_stats(schema: Schema, manifest_no_stats: ManifestFile) -> None:
         ), f"Should read when missing stats for expr: {expr}"
 
 
-def test_not(schema: Schema, manifest: ManifestFile) -> None:
+def test_not_nan(schema: Schema, manifest: ManifestFile) -> None:
```
Comment (rdblue) [inline python/tests/expressions/test_visitors.py:1027]: @Fokko, it looks like these renames were accidental. This is still testing `Not` expressions. I think the suffix `_nan` was added to a lot of these method names unintentionally.

**Review by rdblue — COMMENTED** [review]
The tests look good and I think the implementation is correct. Adding safer NaN handling would be nice, and it looks like some tests were accidentally renamed. Otherwise I think this is about ready to go in.

File: python/pyiceberg/expressions/visitors.py (hunk 19 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1031]: Good one!

File: python/pyiceberg/expressions/visitors.py (hunk 20 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1035]: Always in favor of more Walrus operators

File: python/pyiceberg/expressions/visitors.py (hunk 21 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1073]: Removed!

File: python/pyiceberg/expressions/visitors.py (hunk 11 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1179]: I misinterpreted this comment, let me add those checks for each of the expressions.

Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1179]: I can add those to Java as well.

File: python/pyiceberg/expressions/visitors.py (hunk 22 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py:1204]: Got it, I misread the other comment, thanks!

File: python/pyiceberg/expressions/visitors.py (hunk 12 above)
Comment (Fokko) [reply python/pyiceberg/expressions/visitors.py]: Missed those, added!

File: python/tests/expressions/test_visitors.py (hunk 23 above)
Comment (Fokko) [reply python/tests/expressions/test_visitors.py:1027]: I replicated two suites to Python:
- [TestInclusiveMetricsEvaluator.java](https://github.com/apache/iceberg/blob/master/api/src/test/java/org/apache/iceberg/expressions/TestInclusiveMetricsEvaluator.java)
- [TestMetricsEvaluatorsNaNHandling.java](https://github.com/apache/iceberg/blob/master/api/src/test/java/org/apache/iceberg/expressions/TestMetricsEvaluatorsNaNHandling.java)

The ones coming from `TestMetricsEvaluatorsNaNHandling` have a `_nan` postfix to avoid duplicate naming. But I think we can just combine them.

Comment (rdblue) [reply python/tests/expressions/test_visitors.py]: > The ones coming from TestMetricsEvaluatorsNaNHandling have a _nan postfix to avoid duplicate naming. But I think we can just combine them.

I don't understand why these are NaN tests though. They are identical to the way they were before and don't test anything specific to NaN. Am I missing something?

**Review by rdblue — APPROVED** [review]
I see the NaN tests are using a file with NaN metrics. Looks good now!
